framework/test/jest/unit/specs/modules/chain/synchronizer/fast_chain_switching_mechanism/fast_chain_switching_mechanism.spec.js

```diff
@@ -14,149 +14,866 @@
+				.mockResolvedValue([lastBlock]);
+			// BFT loads blocks from storage and extracts their headers
+			when(storageMock.entities.Block.get)
+				.calledWith(
+					{
+						height_gte: genesisBlockDevnet.height,
+						height_lte: lastBlock.height,
+					},
+					{ limit: null, sort: 'height:asc' },
+				)
+				.mockResolvedValue([genesisBlockDevnet, lastBlock]);
+
+			// Simulate finalized height stored in ChainMeta table is 0
+			when(storageMock.entities.ChainMeta.getKey)
+				.calledWith('BFT.finalizedHeight')
+				.mockResolvedValue(0);
+			jest.spyOn(fastChainSwitchingMechanism, '_queryBlocks');
+			jest.spyOn(fastChainSwitchingMechanism, '_switchChain');
+			jest.spyOn(fastChainSwitchingMechanism, '_validateBlocks');
+			await blocksModule.init();
+			await bftModule.init();
+		});
 
-		let syncMechanism;
+		afterEach(() => {
+			jest.clearAllMocks();
+			// Independently of the correct execution of the mechanisms, `active` property should be always
+			// set to false upon finishing the execution
+			expect(fastChainSwitchingMechanism.active).toBeFalsy();
+		});
+
+		it('should request the common block and try 10 times before giving up. If given up, it should apply a penalty to the peer and restart the mechanism', async () => {
+			// Arrange
+			const storageReturnValue = [
+				{
+					id: genesisBlockDevnet.id,
+				},
+				{
+					id: blocksModule.lastBlock.id,
+				},
+			];
+			when(storageMock.entities.Block.get)
+				.calledWith(
+					{
+						height_in: [2, 1], //  We have lastBlock.height + genesisBlock.height present in DB
+					},
+					{
+						sort: 'height:asc',
+					},
+				)
+				.mockResolvedValue(storageReturnValue);
+			// Simulate peer not sending back a common block
+			when(channelMock.invoke)
+				.calledWith('network:requestFromPeer', {
+					procedure: 'getHighestCommonBlock',
+					peerId: aPeer,
+					data: {
+						ids: storageReturnValue.map(blocks => blocks.id),
+					},
+				})
+				.mockResolvedValue({ data: undefined });
+
+			// Act
+			await fastChainSwitchingMechanism.run(aBlock, aPeer);
 
-		beforeEach(() => {
-			syncMechanism = new FastChainSwitchingMechanism(syncParams);
+			// Assert
+			expect(storageMock.entities.Block.get).toHaveBeenCalledTimes(12); // 10 + 2 from beforeEach hooks
+			expect(channelMock.invoke).toHaveBeenCalledTimes(10);
+			checkApplyPenaltyAndRestartIsCalled(
+				aBlock,
+				aPeer,
+				"Peer didn't return a common block or its height is lower than the finalized height of the chain",
+			);
+			expect(fastChainSwitchingMechanism.active).toBeFalsy();
 		});
 
-		describe('#constructor', () => {
-			it('should create instance of FastChainSwitchingMechanism', async () => {
-				expect(syncMechanism).toBeInstanceOf(FastChainSwitchingMechanism);
+		describe('given that the highest common block is found', () => {
+			it('should apply penalty to the peer and restart syncing mechanisms if the height of the common block is smaller than the finalized height', async () => {
```

Comment:
also,
`describe('when the height of the common block is smaller than the finalized height')` might be better?

Reply:
You mean remove `given that the highest common block is found` describe? I guess we can imply that the common block has been found with `it('should apply penalty to the peer and restart syncing mechanisms if the height of the common block is smaller than the finalized height'`

Reply:
@limiaspasdaniel just understand your perspective here, are you trying to follow the `Given, When, Then` approach? if that's the case `Given would be your hooks, When would be described, Then can be it` what do you think? if this is clear we can change the sentence accordingly.

Reply:
I wasn't strictly following that pattern, especially because we agreed that `it` should always start with `should` but we can come up with a better solution if it is no sufficiently clear right now.